indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java

```diff
@@ -773,4 +790,123 @@ public void setInvalidRowCount(long invalidRowCount)
       this.invalidRowCount = invalidRowCount;
     }
   }
+
+  public static class CombiningIndexGeneratorMapper extends HadoopDruidIndexerMapper<BytesWritable, BytesWritable>
+  {
+    private static final HashFunction hashFunction = Hashing.murmur3_128();
+
+    private AggregatorFactory[] aggregators;
+    private AggregatorFactory[] combiningAggs;
+
+    private Map<Long, IncrementalIndex> combiners = Maps.newHashMap();
+
+    private int counter;
+    private int checkInterval = 20000;
+    private int maxRowsInMemory = 500000;
+
+    @Override
+    protected void setup(Context context) throws IOException, InterruptedException
+    {
+      super.setup(context);
+      aggregators = config.getSchema().getDataSchema().getAggregators();
+      combiningAggs = new AggregatorFactory[aggregators.length];
+      for (int i = 0; i < aggregators.length; ++i) {
+        combiningAggs[i] = aggregators[i].getCombiningFactory();
+      }
+    }
+
+    @Override
+    protected void cleanup(Context context) throws IOException, InterruptedException
+    {
+      for (IncrementalIndex index : combiners.values()) {
+        flush(context, index);
+      }
+      combiners.clear();
+    }
+
+    @Override
+    protected void innerMap(InputRow inputRow, Object value, Context context)
+        throws IOException, InterruptedException
+    {
+      final long timestamp = granularitySpec.getQueryGranularity().truncate(inputRow.getTimestampFromEpoch());
+
+      IncrementalIndex target = combiners.get(timestamp);
+      if (target == null) {
+        combiners.put(timestamp, target = makeIncrementalIndex(timestamp, aggregators, config, null));
```

Comment:
It needs to check whether `inputRow` is instance of `SegmentInputRow`.
It seems that `combiningAggs` should be used instead of `aggregators` in that case.